common/exception/src/exception.rs

```diff
@@ -45,6 +51,139 @@ impl ToString for ErrorCodeBacktrace {
         }
     }
 }
+/// Put Sled's errors wrapper here to avoid circular references
+/// Error returned by a Sled transaction
+#[derive(Debug)]
+pub enum SledTransactionError<T> {
+    Abort(T),
+    /// A storage related error
+    Storage(io::Error),
+}
+
+impl<T: Display> Display for SledTransactionError<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Fmt_Result {
+        match self {
+            Self::Abort(e) => Display::fmt(&e, f),
+            Self::Storage(e) => Display::fmt(&e, f),
+        }
+    }
+}
+
+impl<T: Error + 'static> Error for SledTransactionError<T> {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Self::Abort(e) => Some(e),
+            Self::Storage(e) => Some(e),
+        }
+    }
+}
+
+impl<T> From<TransactionError<T>> for SledTransactionError<T> {
+    fn from(e: TransactionError<T>) -> Self {
+        match e {
+            TransactionError::Abort(e) => Self::Abort(e),
+            TransactionError::Storage(e) => Self::Storage(e.into()),
+        }
+    }
+}
+
+impl<T: Into<io::Error>> From<SledTransactionError<T>> for io::Error {
+    fn from(e: SledTransactionError<T>) -> Self {
+        match e {
+            SledTransactionError::Abort(e) => e.into(),
+            SledTransactionError::Storage(e) => e,
+        }
+    }
+}
+
+/// An error returned from the transaction methods.
+/// Should be returned as-is
+#[derive(Debug)]
+pub enum SledUnabortableTransactionError {
+    #[doc(hidden)]
+    Conflict,
+    /// A regular error
+    Storage(io::Error),
+}
+
+impl Display for SledUnabortableTransactionError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Fmt_Result {
+        match self {
+            Self::Conflict => write!(f, "Transaction conflict"),
+            Self::Storage(e) => Display::fmt(&e, f),
+        }
+    }
+}
+
+impl Error for SledUnabortableTransactionError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Self::Storage(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl From<UnabortableTransactionError> for SledUnabortableTransactionError {
+    fn from(e: UnabortableTransactionError) -> Self {
+        match e {
+            UnabortableTransactionError::Storage(e) => Self::Storage(e.into()),
+            UnabortableTransactionError::Conflict => Self::Conflict,
+        }
+    }
+}
+
+/// An error returned from the transaction closure
+#[derive(Debug)]
+pub enum SledConflictableTransactionError<T> {
```

Comment:
If it is an error returned from a tx closure, it better be a internal error type.
common/exception is a container for error types that are used by more than one crates.